File: platform/middleware/examples/public-rpc.mjs
import { readFile } from "node:fs/promises";
import { createHash } from "node:crypto";
import { parseArgs } from "node:util";
import { PaymentRpc, decodePaymentRequiredHeader, verifyRpcPayment, rpcBatchEvidence, rpcHex } from "../seller/dist/index.js";

const { values } = parseArgs({ options: {
  help: { type: "boolean" }, rpc: { type: "string" }, did: { type: "string" },
  faucet: { type: "string" }, "public-key": { type: "string" }, "claim-key": { type: "string" },
  activity: { type: "string" }, offer: { type: "string" }, authority: { type: "string" }, payer: { type: "string" },
} });
values.rpc ??= process.env.LAYERX_RPC_URL;
values.faucet ??= process.env.LAYERX_FAUCET_URL;
values.did ??= process.env.LAYERX_DID;
if (values.help) {
  process.stdout.write("Usage: node public-rpc.mjs --rpc https://HOST/rpc --did DID [--faucet https://HOST/v1/faucet/claims --public-key HEX32 --claim-key KEY] [--activity SIGNED_HEX_FILE --offer PAYMENT_REQUIRED_HEADER_FILE --authority TRUSTED_AUTHORITY_JSON --payer ACCOUNT_HEX32]\nURLs may come from LAYERX_RPC_URL and LAYERX_FAUCET_URL; DID from LAYERX_DID. Use LAYERX_RPC_TOKEN for submission and LAYERX_FAUCET_TOKEN for faucet authentication. The payment example verifies executed or batched commitment selected by the offer.\n");
} else {
  if (!values.rpc || !values.did) throw new Error("rpc-and-did-required");
  if (new URL(values.rpc).port === "18545" || (values.faucet && new URL(values.faucet).port === "18545")) throw new Error("persistent-host-chain-forbidden");
  const rpc = new PaymentRpc(values.rpc, process.env.LAYERX_RPC_TOKEN ? { authorization: `Bearer ${process.env.LAYERX_RPC_TOKEN}` } : {});
  if (values.faucet) {
    const url = new URL(values.faucet);
    if (url.username || url.password || url.search || url.hash || url.pathname !== "/v1/faucet/claims"
      || (url.protocol !== "https:" && !(url.protocol === "http:" && ["localhost", "127.0.0.1", "[::1]"].includes(url.hostname)))) throw new Error("invalid-faucet-url");
    rpcHex(values["public-key"], 32);
    if (!values["claim-key"] || !process.env.LAYERX_FAUCET_TOKEN) throw new Error("faucet-claim-key-and-token-required");
    const response = await fetch(url, { method: "POST", redirect: "error", signal: AbortSignal.timeout(30_000),
      headers: { "content-type": "application/json", authorization: `Bearer ${process.env.LAYERX_FAUCET_TOKEN}`, "idempotency-key": values["claim-key"] },
      body: JSON.stringify({ did: values.did, public_key: values["public-key"] }) });
    await response.body?.cancel();
    process.stdout.write(`Faucet HTTP status: ${response.status}; confirm the account balance before spending.\n`);
    if (!response.ok) process.exitCode = 1;
  }
  const sequence = await rpc.call("lx_getSequence", [values.did]);
  process.stdout.write(JSON.stringify({ sequence }) + "\n");
  if (values.activity) {
    if (!values.offer || !values.authority || !values.payer) throw new Error("offer-authority-and-payer-required");
    const required = decodePaymentRequiredHeader((await readFile(values.offer, "utf8")).trim());
    const offer = required.accepts.find(v => ["exact", "metered", "subscription"].includes(v.scheme)
      && ["executed", "batched"].includes(v.extra?.layerx?.commitment ?? "executed"));
    if (!offer) throw new Error("executed-or-batched-offer-required");
    const commitment = offer.extra?.layerx?.commitment ?? "executed";
    const canonical = (await readFile(values.activity, "utf8")).trim();
    const activityId = createHash("sha256").update("LXP/v1/activity-id\0").update(rpcHex(canonical)).digest("hex");
    const configured = JSON.parse(await readFile(values.authority, "utf8"));
    const authority = Object.fromEntries(["batchId", "asset", "previousStateRoot", "resultingStateRoot", "sequencerPublicKey"].map(key => [key, rpcHex(configured[key], 32)]));
    const result = await rpc.send(canonical, commitment);
    const commitments = commitment === "executed" ? undefined : { async resolve(receipt, network, requested) {
      if (network !== offer.network || requested !== "batched" || !Number.isSafeInteger(configured.networkId)
        || configured.networkId <= 0 || typeof configured.firstBatchNumber !== "string"
        || !/^(0|[1-9][0-9]*)$/u.test(configured.firstBatchNumber) || typeof configured.lastBatchNumber !== "string"
        || !/^(0|[1-9][0-9]*)$/u.test(configured.lastBatchNumber)) throw new Error("invalid-batch-authority");
      return rpcBatchEvidence(result, activityId, receipt, { networkId: configured.networkId,
        authorization: { sequencerId: rpcHex(configured.sequencerId, 32), publicKey: authority.sequencerPublicKey,
          firstBatchNumber: BigInt(configured.firstBatchNumber), lastBatchNumber: BigInt(configured.lastBatchNumber) } });
    } };
    const payment = await verifyRpcPayment(result, activityId, values.payer, offer, authority, commitments);
    if (payment.kind === "pending") {
      process.stdout.write(JSON.stringify({ state: "pending", activity_id: activityId }) + "\n");
      process.exitCode = 2;
    } else {
      if (offer.scheme !== "exact" && (payment.verification.receipt.moduleId !== 1 || payment.verification.receipt.operation !== 6)) throw new Error("grant-receive-receipt-required");
      const digest = createHash("sha256").update("LXP/v1/merkle-leaf\0").update(payment.canonicalReceipt).digest("hex");
      process.stdout.write(JSON.stringify({ transaction: `lxp:${digest}`, commitment }) + "\n");
    }
  }
}

File: platform/middleware/examples/public_rpc.py
import argparse
import hashlib
import json
import os
import urllib.request
import urllib.parse
from pathlib import Path

from layerx_sdk.x402_http import decode_header, validate_required
from layerx_sdk.x402_rpc import PaymentRpc, rpc_hex, verify_rpc_payment, _NoRedirect
from layerx_sdk.verifier import AuthorizedReceiptBatch


def main():
    parser = argparse.ArgumentParser(
        description="Read public RPC sequence, claim faucet funding, or verify a signed exact payment."
    )
    parser.add_argument("--rpc", default=os.environ.get("LAYERX_RPC_URL"))
    parser.add_argument("--did", default=os.environ.get("LAYERX_DID"))
    parser.add_argument("--faucet", default=os.environ.get("LAYERX_FAUCET_URL"))
    parser.add_argument("--public-key")
    parser.add_argument("--claim-key")
    parser.add_argument("--activity")
    parser.add_argument("--offer")
    parser.add_argument("--authority")
    parser.add_argument("--payer")
    args = parser.parse_args()
    if not args.rpc or not args.did:
        parser.error("RPC URL and DID are required through arguments or environment")
    if urllib.parse.urlsplit(args.rpc).port == 18545 or (
        args.faucet and urllib.parse.urlsplit(args.faucet).port == 18545
    ):
        raise ValueError("persistent-host-chain-forbidden")
    token = os.environ.get("LAYERX_RPC_TOKEN")
    rpc = PaymentRpc(args.rpc, {"Authorization": f"Bearer {token}"} if token else {})
    if args.faucet:
        url = urllib.parse.urlsplit(args.faucet)
        if (
            url.username
            or url.password
            or url.query
            or url.fragment
            or url.path != "/v1/faucet/claims"
            or url.scheme != "https"
            and not (
                url.scheme == "http"
                and url.hostname in ("localhost", "127.0.0.1", "::1")
            )
        ):
            raise ValueError("invalid-faucet-url")
        rpc_hex(args.public_key, 32)
        faucet_token = os.environ.get("LAYERX_FAUCET_TOKEN")
        if not args.claim_key or not faucet_token:
            raise ValueError("faucet-claim-key-and-token-required")
        request = urllib.request.Request(
            args.faucet,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {faucet_token}",
                "Idempotency-Key": args.claim_key,
            },
            data=json.dumps({"did": args.did, "public_key": args.public_key}).encode(),
        )
        with urllib.request.build_opener(_NoRedirect()).open(
            request, timeout=30
        ) as response:
            print(
                f"Faucet HTTP status: {response.status}; confirm the account balance before spending."
            )
    print(json.dumps({"sequence": rpc.call("lx_getSequence", [args.did])}))
    if args.activity:
        if not args.offer or not args.authority or not args.payer:
            raise ValueError("offer-authority-and-payer-required")
        required = validate_required(
            decode_header(Path(args.offer).read_text().strip())
        )
        offer = next(
            v
            for v in required["accepts"]
            if v["scheme"] in ("exact", "metered", "subscription")
            and v.get("extra", {}).get("layerx", {}).get("commitment", "executed")
            == "executed"
        )
        canonical = Path(args.activity).read_text().strip()
        activity = hashlib.sha256(
            b"LXP/v1/activity-id\0" + rpc_hex(canonical)
        ).hexdigest()
        configured = json.loads(Path(args.authority).read_text())
        names = {
            "batch_id": "batchId",
            "asset": "asset",
            "previous_state_root": "previousStateRoot",
            "resulting_state_root": "resultingStateRoot",
            "sequencer_public_key": "sequencerPublicKey",
        }
        authority = AuthorizedReceiptBatch(
            **{key: rpc_hex(configured[name], 32) for key, name in names.items()}
        )
        from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
        from cryptography.exceptions import InvalidSignature

        class Signatures:
            def verify_ed25519(self, public_key, signature, digest):
                try:
                    Ed25519PublicKey.from_public_bytes(public_key).verify(
                        signature, digest
                    )
                    return True
                except (InvalidSignature, ValueError):
                    return False

        result = rpc.send(canonical, "executed")
        verified = verify_rpc_payment(
            result,
            activity,
            args.payer,
            authority,
            Signatures(),
            amount=offer["amount"],
            asset=offer["asset"],
            pay_to=offer["payTo"],
        )
        if verified is None:
            print(json.dumps({"state": "pending", "activity_id": activity}))
            return 2
        if offer["scheme"] != "exact" and (
            verified.receipt.module_id != 1 or verified.receipt.operation != 6
        ):
            raise ValueError("grant-receive-receipt-required")
        digest = hashlib.sha256(
            b"LXP/v1/merkle-leaf\0" + verified.canonical_bytes
        ).hexdigest()
        print(json.dumps({"transaction": "lxp:" + digest, "commitment": "executed"}))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
